integrationtest/backend/tck/src/main/java/org/hibernate/search/integrationtest/backend/tck/analysis/AnalysisBuiltinIT.java

```diff
@@ -18,51 +19,50 @@
 import org.hibernate.search.integrationtest.backend.tck.testsupport.util.SimpleFieldModel;
 import org.hibernate.search.integrationtest.backend.tck.testsupport.util.TckBackendHelper;
-import org.hibernate.search.integrationtest.backend.tck.testsupport.util.rule.SearchSetupHelper;
+import org.hibernate.search.integrationtest.backend.tck.testsupport.util.TckBackendSetupStrategy;
+import org.hibernate.search.integrationtest.backend.tck.testsupport.util.extension.SearchSetupHelper;
 import org.hibernate.search.util.impl.integrationtest.common.assertion.SearchResultAssert;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.SimpleMappedIndex;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.SingleFieldDocumentBuilder;
 import org.hibernate.search.util.impl.integrationtest.mapper.stub.StubMappingScope;
+import org.hibernate.search.util.impl.test.extension.parameterized.ParameterizedClass;
+import org.hibernate.search.util.impl.test.extension.parameterized.ParameterizedSetup;
 
-import org.junit.Before;
-import org.junit.Rule;
-import org.junit.Test;
-import org.junit.runner.RunWith;
-import org.junit.runners.Parameterized;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+import org.junit.jupiter.params.provider.Arguments;
+import org.junit.jupiter.params.provider.MethodSource;
 
 /**
  * Test indexing and searching with built-in analyzer definitions.
  * See {@link AnalyzerNames}.
  */
-@RunWith(Parameterized.class)
-public class AnalysisBuiltinIT {
+@ParameterizedClass
+class AnalysisBuiltinIT {
 
-	@Parameterized.Parameters
-	public static List<SearchSetupHelper> params() {
+	public static List<? extends Arguments> params() {
 		return Arrays.asList(
 				// Test with no analysis configurer whatsoever
-				new SearchSetupHelper( TckBackendHelper::createAnalysisNotConfiguredBackendSetupStrategy ),
+				Arguments.of( (Function<TckBackendHelper,
+						TckBackendSetupStrategy<?>>) TckBackendHelper::createAnalysisNotConfiguredBackendSetupStrategy ),
 				// Test with an analysis configurer that does not override the defaults but defines other analyzers
-				new SearchSetupHelper( TckBackendHelper::createAnalysisCustomBackendSetupStrategy )
+				Arguments.of( (Function<TckBackendHelper,
+						TckBackendSetupStrategy<?>>) TckBackendHelper::createAnalysisCustomBackendSetupStrategy )
 		);
 	}
 
-	@Rule
-	public final SearchSetupHelper setupHelper;
-
+	@RegisterExtension
+	public SearchSetupHelper setupHelper = SearchSetupHelper.create();
 	private final SimpleMappedIndex<IndexBinding> index = SimpleMappedIndex.of( IndexBinding::new );
 
-	public AnalysisBuiltinIT(SearchSetupHelper setupHelper) {
-		this.setupHelper = setupHelper;
-	}
-
-	@Before
-	public void setup() {
-		setupHelper.start().withIndex( index ).setup();
+	@ParameterizedSetup(ParameterizedSetup.Lifecycle.PER_METHOD)
```

Comment:
So, the API is looking great, thanks for that.

One last concern I can raise as I read through the code: there's little reason to combine `ParameterizedSetup.Lifecycle.PER_METHOD` with the `@TestInstance(PER_CLASS)` introduced by `@ParameterizedClass`... And this combination exposes us to bugs in our tests if the `setup` method does not always initialize all fields (leaves them to their default value): then method executions might be affected by the remnants of a previous setup.

Unfortunately, it seems [we can't set the test-instance lifecycle programmatically](https://github.com/junit-team/junit5/blob/71798ca59488362164123053e51f9c28abdcaf16/junit-jupiter-engine/src/main/java/org/junit/jupiter/engine/descriptor/TestInstanceLifecycleUtils.java#L35-L44), so we can't just adapt it automatically based on this "parameterized setup lifecycle".

I wonder if it wouldn't be better to remove the parameter from here, and instead parameterize this using the class-level annotation:

* For "class" setup we'd annotate classes with `@ParameterizedPerClass` (meta-annotated with @TestInstance(PER_CLASS)`.
* For "method" setup we'd annotate classes with `@ParameterizedPerMethod` (not meta-annotated).
* The checks in the filter would verify that classes are annotated with either annotation.

Reply:
hey, I've just tested it and it seems like we can do this:

- leave the `@ParameterizedClass` as is with `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`
- if we need to do `@ParameterizedSetup(ParameterizedSetup.Lifecycle.PER_METHOD)` we add `@TestInstance( TestInstance.Lifecycle.PER_METHOD ) @ParameterizedClass` to the test class. 

The way JUnit first looks for a direct annotation and then checks for a meta-annotations, so we can override the value from the meta `@ParameterizedClass` in specific tests where we need the context to be recreated for each test

Reply:
That's another possibility, the main downside compared to my proposal being that the lifecycle is not explicit for `@ParameterizedClass` and actually defaults to something different than in plain JUnit (PER_CLASS instead of PER_METHOD).

Reply:
Ok, I think this version does it all 😄. There's a `@ParameterizedClass` that doesn't have a `@TestInstance` and which we use to annotate `ParameterizedPerClass` and `ParameterizedPerMethod` (both having `TestInstance`) this way we can still use the `ParameterizedClass` for the lookups where we need to identify tests we are interested in, and also have "shortcuts" for `@ParameterizedClass + @TestInstance` 

I've also moved the test execution into interceptor and dropped the `AfterTestExecutionCallback`

Reply:
These annotations LGTM, feel free to move on to the ReusableOrmSetupHelper stuff :) Or create a separate issue if you intended to work on it separately (maybe you'll have to, if the tests that rely on it use bytecode enhancement?)

I'll mark the PR as draft, please mark it as ready for review when it *is* ready for review :)

Reply:
Well it's already marked as draft, so:

> ~~I'll mark the PR as draft,~~ please mark it as ready for review when it *is* ready for review :)

Reply:
From what I see... the remaining things are:

- [ ] Finish with bytecode enhancement "extension"
- [ ] Replace tests with ReusableOrmSetupHolder
  - [ ] Remove the remaining `org.junit` assumptions/assertions etc
  - [ ] remove extension methods with `*Legacy` suffix
- [ ] Update some of the parameterized tests to use `@ParameterizedClass` instead.
- [ ] Figure out what needs to be fixed in `ActualBackendSetupStrategy#getTestRule()`
- [ ] Check executed test count

If we won't be updating the `main` too much, I can continue adding more to this branch.... otherwise it'll be easier to get what we have in and then continue in a new branch... 🙈

Reply:
> If we won't be updating the `main` too much, I can continue adding more to this branch.... otherwise it'll be easier to get what we have in and then continue in a new branch... 🙈

Personally I'll just merge my pending PRs and then, if there are no new bug reports, I'll probably hold off for a few weeks.